Source/FieldSolver/FiniteDifferenceSolver/EvolveECTRho.cpp

```diff
@@ -0,0 +1,158 @@
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& edge_lengths,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& face_areas,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& ECTRhofield,
+    int lev) {
+
+#if !defined(WARPX_DIM_RZ) and defined(AMREX_USE_EB)
+    if (m_fdtd_algo == MaxwellSolverAlgo::ECT) {
+
+        EvolveRhoCartesianECT(Efield, edge_lengths, face_areas, ECTRhofield, lev);
+
+    }
+#else
+    amrex::ignore_unused(Efield, edge_lengths, face_areas, ECTRhofield, lev);
+#endif
+}
+
+#ifndef WARPX_DIM_RZ
+void FiniteDifferenceSolver::EvolveRhoCartesianECT (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Efield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& edge_lengths,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& face_areas,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& ECTRhofield, const int lev ) {
+#ifdef AMREX_USE_EB
+
+#if !(defined(WARPX_DIM_3D) || defined(WARPX_DIM_XZ))
+    amrex::Abort("EvolveRhoCartesianECT: Embedded Boundaries are only implemented in 2D3V and 3D3V");
+#endif
+
+    amrex::LayoutData<amrex::Real>* cost = WarpX::getCosts(lev);
+
+    // Loop through the grids, and over the tiles within each grid
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+    for (MFIter mfi(*ECTRhofield[0], TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+        if (cost && WarpX::load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::Timers) {
+            amrex::Gpu::synchronize();
+        }
+        Real wt = amrex::second();
+
+        // Extract field data for this grid/tile
+        Array4<Real> const &Ex = Efield[0]->array(mfi);
+        Array4<Real> const &Ey = Efield[1]->array(mfi);
+        Array4<Real> const &Ez = Efield[2]->array(mfi);
+        Array4<Real> const &Rhox = ECTRhofield[0]->array(mfi);
+        Array4<Real> const &Rhoy = ECTRhofield[1]->array(mfi);
+        Array4<Real> const &Rhoz = ECTRhofield[2]->array(mfi);
+        amrex::Array4<amrex::Real> const &lx = edge_lengths[0]->array(mfi);
+        amrex::Array4<amrex::Real> const &ly = edge_lengths[1]->array(mfi);
+        amrex::Array4<amrex::Real> const &lz = edge_lengths[2]->array(mfi);
+        amrex::Array4<amrex::Real> const &Sx = face_areas[0]->array(mfi);
+        amrex::Array4<amrex::Real> const &Sy = face_areas[1]->array(mfi);
+        amrex::Array4<amrex::Real> const &Sz = face_areas[2]->array(mfi);
+
+        // Extract tileboxes for which to loop
+        Box const &trhox = mfi.tilebox(ECTRhofield[0]->ixType().toIntVect());
+        Box const &trhoy = mfi.tilebox(ECTRhofield[1]->ixType().toIntVect());
+        Box const &trhoz = mfi.tilebox(ECTRhofield[2]->ixType().toIntVect());
+
+        amrex::ParallelFor(trhox, trhoy, trhoz,
+
+            [=] AMREX_GPU_DEVICE(int i, int j, int k) {
+                if (Sx(i, j, k) <= 0) return;
+
+#ifndef WARPX_DIM_XZ
```

Comment:
We might need to keep in mind that if we extend this to 1D as well at some point, this `#ifndef` might not do what we want (in this case it would execute this code in 1D too, other than in 3D). Maybe we could add just an inline comment to remind us of this (if what I'm saying is actually correct), as a guardrail for future developments. Similar comments might be needed below as well.

Reply:
In general I am assuming that the ECT method won't be used in 1D. The method itself doesn't really exist in 1D and I don't think we will ever need to go in that direction. On the other hand , it doesn't harm to add an inline comment so I will do it 😄

Reply:
Gosh, that's right. Sorry, my comment was on autopilot ...

Reply:
Always good to check, thanks for the comment 👍